shell/utils/resourceWatcher.js

```diff
@@ -0,0 +1,212 @@
+  STOPPED:    'stopped', // temporarily stopped via message from the socket, a watch should immediately be triggered but the maintenance cycle will pick it up if that doesn't happen.
+  CANCELLED:  'cancelled', // has been flagged to send a stop request to the socket
+  REMOVED:    'removed' // stop request has been sent to the socket or it's been stopped by the socket itself and is now awaiting cleanup by the maintenanceInterval
+};
+
+export default class ResourceWatcher extends Socket {
+  watches = {};
+  maintenanceInterval;
+  connectionMetadata;
+  status = ''
+
+  constructor(url, autoReconnect = true, frameTimeout = null, protocol = null, maxTries = null) {
+    super(url, autoReconnect, frameTimeout, protocol, maxTries);
+    this.baseUrl = self.location.origin + url.replace('subscribe', '');
+
+    this.maintenanceInterval = setInterval(() => {
+      // Every 1 second we:
+      if (this.baseUrl) {
+        this.syncWatches();
+      }
+    }, 1000);
+  }
+
+  setConnectionMetadata(connectionMetadata) {
+    if (!!this.connectionMetadata) {
+      this.connectionMetadata = connectionMetadata;
+    }
+  }
+
+  async watch(watchKey, providedResourceVersion, providedResourceVersionTime, providedKeyParts = {}, providedSkipResourceVersion) {
+    const {
+      PENDING, REQUESTED, WATCHING, REFRESHING
+    } = WATCHSTATUSES;
+    const {
+      resourceType: providedResourceType,
+      id: providedId,
+      namespace: providedNamespace,
+      selector: providedSelector
+    } = providedKeyParts;
+    const resourceType = providedResourceType || this.watches?.[watchKey]?.resourceType;
+    const id = providedId || this.watches?.[watchKey]?.id;
+    const namespace = providedNamespace || this.watches?.[watchKey]?.namespace;
+    const selector = providedSelector || this.watches?.[watchKey]?.selector;
+    const resourceStatus = this.watches?.[watchKey]?.status;
+    const skipResourceVersion = this.watches?.[watchKey]?.skipResourceVersion || providedSkipResourceVersion;
+    const resourceUrl = this.baseUrl + resourceType;
+    const limitedResourceUrl = addParam(resourceUrl, 'limit', 1);
+    const opt = {
+      method:  'get',
+      headers: { accept: 'application/json' }
+    };
+    const watchObject = {
+      resourceType,
+      id,
+      namespace,
+      selector
+    };
+
+    let resourceVersionTime = providedResourceVersionTime || this.watches?.[watchKey]?.resourceVersionTime;
+    let resourceVersion = providedResourceVersion || this.watches?.[watchKey]?.resourceVersion;
+
+    if (!skipResourceVersion && (!resourceVersion || Date.now() - resourceVersionTime > 300000)) { // 300000ms is 5minutes
+      await fetch(limitedResourceUrl, opt)
+        .then((res) => {
+          if (!res.ok) {
+            this.watches[watchKey].error = res.json();
+            console.warn(`Resource error retrieving resourceVersion`, resourceType, ':', res.json()); // eslint-disable-line no-console
+          }
+
+          return res.json();
```

Comment:
Should we continue if we've failed to get the revision, won't resourceVersionTime be tracked but resourceVersionTime supplied?

Reply:
there's some resources that simply don't return a revision (nodemetrics being one of them I think) so I don't want the socket to just die on the vine here if that's the case, especially if there wasn't an error. I added a flag to the watch to keep track of that case though so we can use the error to know how the fetch failed.